### PR TITLE
Add pootle_log.Log

### PR DESCRIPTION
**Pull request by phlax**
base log class for retrieving LogEvents

### CODE REVIEW

File: pootle/apps/pootle_log/utils.py [hunk 1]
```diff
@@ -18,3 +24,141 @@ def __init__(self, unit, user, timestamp, action, value,
...
+    def created_unit_model(self):
+        return UnitSource.objects
+
+    @property
+    def suggestion_model(self):
+        return Suggestion.objects
+
+    @property
+    def submission_model(self):
+        return Submission.objects
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: This and the ones above are not models, but managers.

File: pootle/apps/pootle_log/utils.py [hunk 2]
```diff
@@ -18,3 +24,141 @@ def __init__(self, unit, user, timestamp, action, value,
...
+
+    def filter_path(self, qs, path, field="unit__store__pootle_path"):
+        return (
+            qs.filter(**{"%s__startswith" % field: path})
+            if path is not None
+            else qs)
+
+    def filter_store(self, qs, store, field="unit__store_id"):
+        return (
+            qs.filter(**{"%s" % field: store})
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: Why not just simpler:

```python
            qs.filter(**{field: store})
```

File: pootle/apps/pootle_log/utils.py [hunk 3]
```diff
@@ -18,3 +24,141 @@ def __init__(self, unit, user, timestamp, action, value,
...
+            if store is not None
+            else qs)
+
+    def filter_times(self, qs, start, end, field="creation_time"):
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: Shouldn't the default values for `start` and `end` be explicit here?

```python
    def filter_times(self, qs, start=None, end=None, field="creation_time"):
```

Comment (unho) [inline pootle/apps/pootle_log/utils.py]: Same for other methods.

File: pootle/apps/pootle_log/utils.py [hunk 4]
```diff
@@ -18,3 +24,141 @@ def __init__(self, unit, user, timestamp, action, value,
...
+
+    def filtered_submissions(self, **kwargs):
+        submissions = self.submission_model
+        submissions = self.filter_store(
+            self.submission_model,
+            kwargs.get("store"))
+        submissions = (
+            self.filter_user(
+                submissions,
+                kwargs.get("user")))
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: Nit: This fits in a single line. I guess it was kept like this to ease reading.

File: pootle/apps/pootle_log/utils.py (hunk 2 above)
Comment (phlax) [inline pootle/apps/pootle_log/utils.py]: yep

File: pootle/apps/pootle_log/utils.py (hunk 3 above)
Comment (phlax) [inline pootle/apps/pootle_log/utils.py]: yep

File: pootle/apps/pootle_log/utils.py (hunk 4 above)
Comment (phlax) [inline pootle/apps/pootle_log/utils.py]: quite a lot of refactoring - so probs it was longer previously

File: pootle/apps/pootle_log/utils.py [hunk 5]
```diff
@@ -18,3 +24,141 @@ def __init__(self, unit, user, timestamp, action, value,
...
+    @property
+    def created_unit_model(self):
+        return UnitSource.objects.select_related("unit", "created_by")
+
+    @property
+    def suggestion_model(self):
+        return Suggestion.objects.select_related("unit", "user", "reviewed_by")
+
+    @property
+    def submission_model(self):
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: These are still managers.

File: pootle/apps/pootle_log/utils.py (hunk 3 above)
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: I think the same for `user` and other params used for filtering.

File: pootle/apps/pootle_log/utils.py (hunk 4 above)
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: I am ok with keeping this as is.

File: pootle/apps/pootle_log/utils.py (hunk 5 above)
Comment (phlax) [inline pootle/apps/pootle_log/utils.py]: i think they might actually be qs as they have select_related

Comment (unho) [inline pootle/apps/pootle_log/utils.py]: You are right. But definitely not models ;)

**Comment by unho** [comment]
Just wondering if there should be any filtering by unit as well.

**Comment by phlax** [comment]
> Just wondering if there should be any filtering by unit as well.

i also wondered, but decided against on the basis that you should use UnitLog in that case

File: pootle/apps/pootle_log/utils.py [hunk 6]
```diff
@@ -18,3 +24,145 @@ def __init__(self, unit, user, timestamp, action, value,
...
+        return Submission.objects.select_related("unit", "submitter")
+
+    @cached_property
+    def event(self):
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: Nit: rename to `event_model`?

File: pootle/apps/pootle_log/utils.py [hunk 7]
```diff
@@ -18,3 +24,145 @@ def __init__(self, unit, user, timestamp, action, value,
...
+            for n
+            in ["SOURCE", "TARGET", "STATE", "COMMENT", "CHECK"]}
+
+    @cached_property
+    def subtypes(self):
+        return {
+            getattr(SubmissionTypes, n): n.lower()
+            for n in ["WEB", "UPLOAD", "SYSTEM"]}
+
+    def filter_path(self, qs, path, field="unit__store__pootle_path"):
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: `path=None`

File: pootle/apps/pootle_log/utils.py (hunk 6 above)
Comment (phlax) [inline pootle/apps/pootle_log/utils.py]: thats really not a model!

im gonna rename the others

File: pootle/apps/pootle_log/utils.py [hunk 8]
```diff
@@ -143,3 +143,56 @@ def filtered_created_units(self, **kwargs):
...
+                 and (not kwargs.get("end")
+                      or (suggestion.creation_time < kwargs.get("end")))
+                 and(not kwargs.get("user")
+                     or (suggestion.user == kwargs.get("user")))))
+            review_event = (
+                (not kwargs.get("start")
+                 or (suggestion.review_time >= kwargs.get("start"))
+                 and (not kwargs.get("end")
+                      or (suggestion.review_time < kwargs.get("end")))
+                 and(not kwargs.get("user")
```
Comment (unho) [inline pootle/apps/pootle_log/utils.py]: no space before parentheses

**Comment by unho** [comment]
lgtm

**Pull request merged**